trieste/models/gpflow/inducing_point_selectors.py

```diff
@@ -0,0 +1,224 @@
+        :raise tf.errors.InvalidArgumentError: If ``dataset`` is empty.
+        """
+
+        tf.debugging.Assert(len(dataset.query_points) is not None, [])
+
+        N = tf.shape(dataset.query_points)[0]  # training data size
+        shuffled_query_points = tf.random.shuffle(dataset.query_points)  # [N, d]
+        sub_sample = shuffled_query_points[: tf.math.minimum(N, M), :]
+
+        if N < M:  # if fewer data than inducing points then sample remaining uniformly
+            uniform_sampler = UniformInducingPointSelector(self._search_space)
+            uniform_sample = uniform_sampler._recalculate_inducing_points(
+                M - N, model, dataset
+            )  # [M-N, d]
+            sub_sample = tf.concat([sub_sample, uniform_sample], 0)  # [M, d]
+        return sub_sample  # [M, d]
+
+
+class KMeansInducingPointSelector(InducingPointSelector[SupportsGetInducingVariables]):
+    """
+    An :class:`InducingPointSelector` that chooses points as centroids of a K-means clustering
+    of the training data.
+    """
+
+    def _recalculate_inducing_points(
+        self, M: int, model: SupportsGetInducingVariables, dataset: Dataset
+    ) -> TensorType:
+        """
+        Calculate `M` centroids from a K-means clustering of the training data.
+
+        If the clustering returns fewer than `M` centroids or if we have fewer than `M` training
+        data, then we fill the remaining points with random samples across the search space.
+
+        :param M: Desired number of inducing points.
+        :param model: The sparse model.
+        :param dataset: The data from the observer. Must be populated.
+        :return: The new updated inducing points.
+        :raise tf.errors.InvalidArgumentError: If ``dataset`` is empty.
+        """
+
+        tf.debugging.Assert(dataset is not None, [])
+
+        query_points = dataset.query_points  # [N, d]
+        N = tf.shape(query_points)[0]
+
+        shuffled_query_points = tf.random.shuffle(query_points)  # [N, d]
+        query_points_stds = tf.math.reduce_std(shuffled_query_points, 0)  # [d]
+
+        if (
+            tf.math.count_nonzero(query_points_stds, dtype=N.dtype) == N
+        ):  # standardize if all stds non zero
+            normalize = True
+            shuffled_query_points = shuffled_query_points / query_points_stds  # [N, d]
+        else:
+            normalize = False
+
+        centroids, _ = kmeans(shuffled_query_points, int(tf.math.minimum(M, N)))  # [C, d]
+
+        if len(centroids) < M:  # choose remaining points as random samples
+            uniform_sampler = UniformInducingPointSelector(self._search_space)
+            extra_centroids = uniform_sampler._recalculate_inducing_points(  # [M-C, d]
+                M - len(centroids), model, dataset
+            )
+            extra_centroids = extra_centroids / query_points_stds  # remember to standardize
+            centroids = tf.concat([centroids, extra_centroids], axis=0)  # [M, d]
+
```

Comment:
not sure how inducing points enter the training, but does the order matter? just wodnering if it would be good to shuffle them after this concatenation... (same thing for randomsubsampling one)

Reply:
order doesnt matter. We shuffle  them anyway for the ELBO